### PR TITLE
Refactor reasoner oracle test rule partitioning

### PR DESCRIPTION
**Pull request by flyingsilverfin**
## What is the goal of this PR?

Simplify and restructure the rule partitioning (aka Stratification in the literature) approach implemented in the Oracle test framework. We improve readability by creating smaller methods and more canonical state.

## What are the changes implemented in this PR?

* deleted derived/unnecessary state
* apply new naming to help the reader distinguish the state
* refactor revealed common code paths that could be deduplicated

### CODE REVIEW

File: test/behaviour/reasoner/verification/ForwardChainingMaterialiser.java [hunk 1]
```diff
@@ -118,52 +118,48 @@ Optional<Materialisation> conditionMaterialisations(com.vaticle.typedb.core.logi
...
-        // Computes the stratification of rules required by stratified-negation.
+        // Compute the stratification of rules required by stratified-negation.
         if (rulePartitions == null) {
             rulePartitions = new LinkedList<>();
-            Set<Rule> inLowerPartitions = new HashSet<>(); // Tracks nodes already added to some partition.
             for (Rule rule : rules.values()) {
-                if (!inLowerPartitions.contains(rule)) {
-                    Set<Rule> inLowerOrSamePartition = new HashSet<>();
-                    partitioningDFS(rule, inLowerPartitions, inLowerOrSamePartition);
-                    updatePartitions(inLowerPartitions, inLowerOrSamePartition);
```
Comment (flyingsilverfin) [inline test/behaviour/reasoner/verification/ForwardChainingMaterialiser.java:129]: push derived logic one method down, which also makes it re-usable

File: test/behaviour/reasoner/verification/ForwardChainingMaterialiser.java [hunk 2]
```diff
@@ -118,52 +118,48 @@ Optional<Materialisation> conditionMaterialisations(com.vaticle.typedb.core.logi
...
-            for (Rule dependency : at.negatedDependencies()) {
-                if (!inLowerPartitions.contains(dependency)) {
-                    Set<Rule> inLowerOrSamePartitionAsDependency = new HashSet<>();
-                    partitioningDFS(dependency, inLowerPartitions, inLowerOrSamePartitionAsDependency);
-                    updatePartitions(inLowerPartitions, inLowerOrSamePartitionAsDependency);
-                }
-            }
+    private boolean inPartition(Rule rule) {
+        return iterate(rulePartitions).anyMatch(partition -> partition.contains(rule));
+    }
```
Comment (flyingsilverfin) [inline test/behaviour/reasoner/verification/ForwardChainingMaterialiser.java:133]: replace separate `Set` with looping over the previously created partitions

File: test/behaviour/reasoner/verification/ForwardChainingMaterialiser.java [hunk 3]
```diff
@@ -118,52 +118,48 @@ Optional<Materialisation> conditionMaterialisations(com.vaticle.typedb.core.logi
...
+        extractPartition(positivelyReachable);
+    }
+
+    private void partitionDependencies(Rule rule, Set<Rule> positivelyReachable) {
+        // Negated dependencies of a rule are strictly in a lower partition.
+        // Positive dependencies of a rule are in the current partition or a lower one.
+        rule.negatedDependencies().forEach(negatedDependency -> {
+            if (!inPartition(negatedDependency)) computePartitionsFrom(negatedDependency);
+        });
+        rule.unnegatedDependencies().forEach(dependency -> expandPartition(dependency, positivelyReachable));
```
Comment (flyingsilverfin) [inline test/behaviour/reasoner/verification/ForwardChainingMaterialiser.java:148]: simplified the dependency partitioning - now it's very clear that we have two separate behaviours for negated dependencies and positive dependencies

**Comment by krishnangovindraj** [comment]
This refactor describes the code in the properties of the nodes (connectedness) rather than the final result which the properties tell you (which stratum they are in). It also makes explicit the difference between steps into lower stratum (computePartitionFrom) and expanding the current stratum (partitionDependencies).
This makes it easier to follow what the code is doing, which we know will lead to the answer as post-condition.
The earlier approach was harder to decipher as it talked in-terms of the post-condition before it was achieved, leading to some confusion.

**Pull request merged**